### PR TITLE
New Inbox design

### LINKED ISSUE #1727: Inbox cell redesign
Teaching myself Sketch.app and using the inbox cell as a starting point. I'm trying to accomplish a few things:

- Faster interactions (tap button instead of hidden swipe)
- More, balanced whitespace
- Better information hierarchy

cc @orta you've always had great insight!

screen shot 2018-04-14 at 6 02 17 pm

### PR DESCRIPTION
**Pull request by rnystrom**
Fixes #1727 

[Video](https://twitter.com/_ryannystrom/status/1005853590118944769)

Things included in this PR:

- Subscribe/unsubscribe working
- "More" button with owner/repo actions
- Tap checkmark to read, fades cell, refresh to remove cell
- Use `IGListKit+Swift` and `struct` models
  - Discovered a bug w/ diffing structs not using typename. Will fix upstream and update in another PR.
- Shorthand timestamps to leave more room for owner/repo#123 format
- Removed cell swiping, much simpler UX (idea from Twitter.app)
- Moved some code into shared functions

### CODE REVIEW

File: Classes/Notifications/NotificationCell2.swift [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+    private let dateLabel = ShowMoreDetailsLabel()
+    private let detailsLabel = UILabel()
+    private let textView = StyledTextView()
+    private let stackView = UIStackView()
+    private let commentButton = UIButton()
+    private let readButton = UIButton()
+    private let watchButton = UIButton()
+    private let moreButton = UIButton()
+
+    override init(frame: CGRect) {
```
Comment (jdisho) [inline Classes/Notifications/NotificationCell2.swift]: I think the `init` is too large. I would split it up in smaller `functions` and call these `functions` in the `init`. 
e.g:

```swift

override init(frame: CGRect) {
   ...
   
   configureCommentButton()
   ...
}

private func configureCommentButton() {
    contentView.addSubview(commentButton)

    commentButton.titleLabel?.font = font
    commentButton.isUserInteractionEnabled = false
    commentButton.tintColor = grey
    commentButton.setTitleColor(grey, for: .normal)
    commentButton.contentEdgeInsets = UIEdgeInsets(top: 0, left: 0, bottom: -2, right: 0)
    commentButton.titleEdgeInsets = UIEdgeInsets(top: -4, left: 2, bottom: 0, right: 0)
    commentButton.setImage(UIImage(named: "comment-small")?.withRenderingMode(.alwaysTemplate), for: .normal)
    commentButton.contentHorizontalAlignment = .left

    commentButton.snp.makeConstraints { make in
        make.width.equalTo(actionsHeight)
    }
}
```

File: Local Pods/GitHubAPI/GitHubAPI/V3SubscribeThreadRequest.swift [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+import Foundation
+
+public struct V3SubscribeThreadRequest: V3Request {
+    public typealias ResponseType = V3StatusCodeResponse<V3StatusCode200>
+    public var pathComponents: [String] {
+        return ["notifications", "threads", id, "subscription"]
+    }
+    public var method: HTTPMethod { return .put }
+    public var parameters: [String : Any]? {
+        return [
```
Comment (jdisho) [inline Local Pods/GitHubAPI/GitHubAPI/V3SubscribeThreadRequest.swift]: Why not doing the same as for the `pathComponents` ?

File: Classes/View Controllers/BaseListViewController2.swift [hunk 3]
```diff
@@ -13,15 +13,16 @@ protocol BaseListViewController2DataSource: class {
...
+class BaseListViewController2<PageType: CustomStringConvertible>: UIViewController,
 ListSwiftAdapterDataSource,
-FeedDelegate {
+FeedDelegate,
+LoadMoreSectionController2Delegate {
 
     private let emptyErrorMessage: String
     public weak var dataSource: BaseListViewController2DataSource?
 
     public private(set) lazy var feed: Feed = { Feed(viewController: self, delegate: self) }()
```
Comment (jdisho) [inline Classes/View Controllers/BaseListViewController2.swift]: Why `public` here and for the `dataSource` 2 lines above? Isn't `internal` enough?

File: Classes/Notifications/NotificationCell2.swift (hunk 1 above)
Comment (rnystrom) [reply Classes/Notifications/NotificationCell2.swift]: Unless the functions are going to be used again, I think this gets too confusing and is error prone.

What happens if one of these configs depends on another view being added first? Now they implicitly depend on each other w/out any way of knowing the order of operations.

Putting them all inline makes it extremely simple, easy to understand, and shows that this setup isn't ever called again.

File: Classes/Models/Repository.swift [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+//  Freetime
+//
+//  Created by Ryan Nystrom on 6/9/18.
+//  Copyright © 2018 Ryan Nystrom. All rights reserved.
+//
+
+import Foundation
+
+struct Repository {
+    var owner: String
```
[the remark this answers is not in the document]
Comment (rnystrom) [reply Classes/Models/Repository.swift]: It's a `struct` now, so `let` locks the model down for us. Check out how model is updated in the `FlatCache`.

File: Classes/Notifications/Notification+NotificationViewModel.swift [hunk 5]
```diff
@@ -45,3 +45,43 @@ func CreateViewModels(
...
+    contentSizeCategory: UIContentSizeCategory,
+    v3notifications: [V3Notification]) -> [NotificationViewModel2] {
+    var models = [NotificationViewModel2]()
+
+    for notification in v3notifications {
+        guard let type = NotificationType(rawValue: notification.subject.type.rawValue),
+            let identifier = notification.subject.identifier
+            else { continue }
+
+        let modelIdentifier: NotificationViewModel2.Identifier
```
[the remark this answers is not in the document]
Comment (rnystrom) [reply Classes/Notifications/Notification+NotificationViewModel.swift]: I know, the naming here sucks. Will come up w/ a better schema. Right now there's

- `ident` (number, hash, release #)
- `v3id` (v3 API id)
- `id` (for `FlatCache`)
- `identifier` (for `ListSwiftDiffable`)

File: Classes/Notifications/NotificationCell2.swift [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+final class NotificationCell2: SelectableCell {
+
+    public static let inset = UIEdgeInsets(
+        top: NotificationCell2.topInset + NotificationCell2.headerHeight + Styles.Sizes.rowSpacing,
+        left: Styles.Sizes.icon.width + 2*Styles.Sizes.columnSpacing,
+        bottom: NotificationCell2.actionsHeight,
+        right: Styles.Sizes.gutter
+    )
+    public static let topInset = Styles.Sizes.rowSpacing
+    public static let headerHeight = ceil(Styles.Text.secondary.preferredFont.lineHeight)
```
[the remark this answers is not in the document]
Comment (rnystrom) [reply Classes/Notifications/NotificationCell2.swift]: It's a `UIFont` lineHeight, want to snap to a pixel

File: Classes/Notifications/NotificationsViewController2.swift [hunk 7]
```diff
@@ -0,0 +1,284 @@
...
+            )
+            item.accessibilityLabel = NSLocalizedString("More options", comment: "")
+            navigationItem.leftBarButtonItem = item
+        case .repo, .all: break
+        }
+
+        navigationController?.tabBarItem.badgeColor = Styles.Colors.Red.medium.color
+    }
+
+    override func fetch(page: Int?) {
```
[the remark this answers is not in the document]
Comment (rnystrom) [reply Classes/Notifications/NotificationsViewController2.swift]: Override from `BaseListViewController2` w/ generic type

File: Classes/Notifications/NotificationsViewController2.swift [hunk 8]
```diff
@@ -0,0 +1,284 @@
...
+        switch result {
+        case .success(let notifications, let next):
+            var ids = [String]()
+            notifications.forEach {
+                modelController.githubClient.cache.add(listener: self, value: $0)
+                ids.append($0.id)
+            }
+            rebuildAndUpdate(ids: ids, append: append, page: next, animated: animated)
+        case .error:
+            error(animated: trueUnlessReduceMotionEnabled)
```
[the remark this answers is not in the document]
Comment (rnystrom) [reply Classes/Notifications/NotificationsViewController2.swift]: Show's empty error message

File: Local Pods/GitHubAPI/GitHubAPI/V3SubscribeThreadRequest.swift (hunk 2 above)
Comment (rnystrom) [reply Local Pods/GitHubAPI/GitHubAPI/V3SubscribeThreadRequest.swift]: What do you mean?

File: Classes/View Controllers/BaseListViewController2.swift (hunk 3 above)
Comment (rnystrom) [reply Classes/View Controllers/BaseListViewController2.swift]: Ya

File: Local Pods/GitHubAPI/GitHubAPI/V3SubscribeThreadRequest.swift (hunk 2 above)
Comment (jdisho) [reply Local Pods/GitHubAPI/GitHubAPI/V3SubscribeThreadRequest.swift]: Doing it like this
```swift 
return ["ignored": ignore ? "true" : "false"]
```

**Comment by rnystrom** [comment]
TODO: scrolling is choppy, we might be missing caches

**Comment by rnystrom** [comment]
Bug w/ marking read, scrolling off screen, then scrolling back.

**Comment by rnystrom** [comment]
Good point on the alpha. Could definitely use some love. The comment “button” just displays the thread. Maybe tapping it displays the thread and focuses the keyboard?

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by BasThomas** [comment]
Hm, that does not feel too intuitive to me, nor would it be too logical to focus on the keyboard as you wouldn't know the contents of the latest update(s) to the thread.

**Comment by rnystrom** [comment]
FWIW that’s what Twitter does.

@BasThomas is there something you have in mind? Are you expecting a certain behavior? Is this something that repeatedly trips you up or did you get the hang of it?

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by BasThomas** [comment]
Yeah but Twitter has the cells filled with the content, whereas here we have the content in the detail.

I did get the hang of it, although the swiping muscle memory is still there.

If we can fix the alpha (don't alpha the still accessible buttons), then I think its 👍

**Comment by rnystrom** [comment]
@BasThomas PRs welcome 😁

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Pull request merged**